extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/secret/KsmClient.java

```diff
@@ -250,13 +290,26 @@ private void validateCache() throws GuacamoleException {
                 String hostname = recordService.getHostname(record);
                 addRecordForHost(record, hostname);
 
-                // Store based on username ONLY if no hostname (will otherwise
+                // ... and domain
+                String domain = recordService.getDomain(record);
+                addRecordForDomain(record, domain);
+
+                // Get the username off of the record
+                String username = recordService.getUsername(record);
+
+                // If domain matching is not enabled for user records,
+                // explicitly set all domains to null to allow matching
+                // on username only
+                if (!confService.getMatchUserRecordsByDomain())
+                    domain = null;
```

Comment:
I recall from a previous review that the `forEach()` changed to an iterator specifically to support checked exceptions. Is this is the only case where that exception is thrown?

If so, then I think we should avoid both the repeated parsing of the config value and the limitations of Java's lambdas by just grabbing this config value once before we enter the `forEach()`. It doesn't need to be re-grabbed for each record. ;)

Reply:
Well, that's not the only config parsing that's going on now - `getUsername` and `getDomain` now also parse the config every time they're called - a consequence of the refactor where that logic got shoved into the record service. I guess I could read the config once on service startup or something, or we could just undo that refactor and read all the configs once at the start of the loop here.

Reply:
I know that we read from a config file to parse the config so it's not going to change but it still feels weird to pull the config once at service creation time when there's no actual guarantee in the API that the config will never change.

Reply:
Very true, but I don't mean at creation time. I was thinking something like:

```java
final boolean matchUserRecordsByDomain = confService.getMatchUserRecordsByDomain();
records.forEach(record -> {
    ...
});
```

The config would then be refreshed each time the cache is refreshed, rather than once per record. But:

> Well, that's not the only config parsing that's going on now - `getUsername` and `getDomain` now also parse the config every time they're called ...

That's a really good point. It sounds like the minor performance hit and loss of nifty lambda isn't worth the added complexity.

Reply:
Sorry, I'm a little unclear exactly what you mean by "It sounds like the minor performance hit and loss of nifty lambda isn't worth the added complexity.". 

Are you suggesting I get rid of the refactor that moved the conf parsing into the record service?

I think that would be my preference - to just read all the configs once before this here `forEach` and get rid of all the exception-throwing stuff in the domain service.

We could go back to using a lambda that way too.